### PR TITLE
Fix various issues with user auth + refactoring + tweaks

### PR DESCRIPTION
**Pull request by jits**
**Note:** this requires a database reset – the easiest way is to delete your existing db and run `bin/setup`.

**Highlights:**
- New feature: don't allow user to complete sign in if email is not yet verified (instead we show an alert message).
- Add UI for flash messages.
- Separate out controller auth logic into a separate controller concern + add specs.
- `AuthService` -> `FirebaseAuthService` and is now just focused on handling the Firebase token.
- Only bootstrap the `AuthApp` VueJS component on the Sign In page (note: all JS is still served as one pack to the client).
- Refactor the `SignIn` VueJS component.
- Don't register the Firebase app and Firebase Auth UI multiple times on the client (prevents client side errors from occuring when navigating back to the Sign In page).
- Redo the `User` model to add UUID primary key + required constraints + unique indexes.
- New feature: track when we last see a user (via a `last_seen_at` that's set every time they sign in or we recognise an existing user session).
- Ensure the Travis build uses Postgres v10.
- Add more caching to the Travis build.
- Add env vars for review apps.

**Known issues remaining:**
- The public key for the JWT verification needs to be cached in Redis.
- Need ability to manually send the email verification again, on user's request.
- If your email is not verified then every time you sign in we resend the email verification.
- The flow from "redirect because sign in is required" -> "user signs in and hits callback" -> "callback says that email is not verified" is not quite working, and is a bit janky – needs proper fixing.
- Clicking on the sign in nav item (in the header) doesn't remember the page you started on – we could make this open a dialog in the future to get around this.
- Potentially handle scenario where:
  - A user account is deleted on firebase
  - Then the same email is signed into through the site.
  - Then we get a validation error saying that the email has been taken (since it's a new `firebase_id`, but an existing user account with that email in the db).
  - NOTE: maybe this is a case of only deleting user accounts from the website (which ensures the db is cleaned up too).
- Improved error handling from server --> client is required for cases like the above (the /auth/callback endpoint) – right now it will respond with an HTML error page when it errors, and the XHR call won't do anything about it.

### CODE REVIEW

**Comment by jits** [comment]
Looking into Firebase UI a bit more, it looks like we need to properly decouple the logic that happens after we get a successful sign in (otherwise we get some odd timing issues). I will attempt to do this using Vue's event system.

**Comment by jits** [comment]
On second thoughts, I can do that in a separate PR. Just so we can get this in sooner than later.

File: db/migrate/20190128014059_create_users.rb [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
-    create_table :users do |t|
-      t.string :display_name
-      t.string :uid
-      t.string :email
+    create_table :users, id: :uuid do |t|
+      t.string :display_name, null: false
+      t.string :firebase_id, null: false
+      t.string :email, null: false
+      t.boolean :email_verified, default: false
+      t.datetime :last_seen_at
```
Comment (fahadiqbal1) [inline db/migrate/20190128014059_create_users.rb]: When is this updated? Also is this necessary?

File: config/locales/en.yml [hunk 2]
```diff
@@ -5,7 +5,7 @@ en:
     language: language
   site-header:
     home: Home
-    sign-in: Sign In
+    sign-in: Create Account / Sign In
```
Comment (fahadiqbal1) [inline config/locales/en.yml]: Maybe this is a discussion with the rest of the group, but maybe we go with "Register"?

File: app/webpack/packs/application.js [hunk 3]
```diff
@@ -10,11 +10,12 @@ import TurbolinksAdapter from 'vue-turbolinks';
...
 
 Rails.start();
-Vue.use(TurbolinksAdapter);
 Turbolinks.start();
 
+Vue.use(TurbolinksAdapter);
+
 document.addEventListener('turbolinks:load', () => {
-  const element = document.getElementById('application');
+  const element = document.getElementById('sign_in');
```
Comment (fahadiqbal1) [inline app/webpack/packs/application.js]: So are we going to have different Vue app instances for different features? In other projects, I have just gone with one global project called `application` that only loads components if the page has it defined.

File: db/migrate/20190127000000_enable_pgcrypto.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class EnablePgcrypto < ActiveRecord::Migration[5.2]
+  def change
+    enable_extension 'pgcrypto'
```
Comment (fahadiqbal1) [inline db/migrate/20190127000000_enable_pgcrypto.rb]: A superuser is required to enable this. This should be done by an admin and not by the application's db user, no?

File: app/webpack/packs/application.js (hunk 3 above)
Comment (jits) [reply app/webpack/packs/application.js]: Yeah, I know what you mean - I usually do the same. In this particular case I expect (currently) we'll only have a handful of Vue components, and only active on certain pages, so it seemed more relevant to only bootstrap in those cases.

File: config/locales/en.yml (hunk 2 above)
Comment (jits) [reply config/locales/en.yml]: Good idea. Will change this. And then can change later, if needed, once we've had a chance to discuss with team.

I think the whole header needs a bit of a rethink - was going to add the user's display name, but there's no obvious place. Maybe as a user dropdown (will try to use this actually).

File: db/migrate/20190127000000_enable_pgcrypto.rb (hunk 4 above)
Comment (jits) [reply db/migrate/20190127000000_enable_pgcrypto.rb]: I don't think so? We expect dbs to have super user access anyway? FWIW, this is the recommended way to do stuff like this.

File: db/migrate/20190128014059_create_users.rb (hunk 1 above)
Comment (jits) [reply db/migrate/20190128014059_create_users.rb]: Check out the `Authentication` concern below.

It's useful to know which users are inactive - e.g. to purge inactive user accounts, or to re-engage.

File: config/locales/en.yml (hunk 2 above)
Comment (fahadiqbal1) [reply config/locales/en.yml]: Actually, I add that second dropdown in the PR im working on, right now I have it as:

* Username
  * My Profile
  * My Subscriptions
  * Sign Out

File: db/migrate/20190128014059_create_users.rb (hunk 1 above)
Comment (fahadiqbal1) [reply db/migrate/20190128014059_create_users.rb]: Shoot I just glanced right past it!

File: config/locales/en.yml (hunk 2 above)
Comment (jits) [reply config/locales/en.yml]: Sounds good! I'll need this for my subscriptions work – could you add that as a commit to this branch and push it up?

File: db/migrate/20190128014059_create_users.rb (hunk 1 above)
Comment (jits) [reply db/migrate/20190128014059_create_users.rb]: That was the main file in this PR! 😄

File: config/locales/en.yml (hunk 2 above)
Comment (fahadiqbal1) [reply config/locales/en.yml]: Sure, can do that right now actually

Comment (fahadiqbal1) [reply config/locales/en.yml]: Done 😃

Comment (jits) [reply config/locales/en.yml]: Nice one!

**Comment by jits** [comment]
Looks like the production build is not working (i.e. the Sign In widget is not loading on the review app) – I'll take a look at this in a separate PR, and merge this one in to unblock upcoming work.

/cc @fahadiqbal1

**Pull request merged**